create User Api

JazminGonzalez-Rivero opened the pull request:
closes #377

jessesuen commented on server/users/users.proto:
```diff
@@ -0,0 +1,37 @@
+import "google/api/annotations.proto";
+
+
+message Body {
```
Theres no need for a Body message, just put these fields in theUsersPasswordRequest. In the grpc option you can put `*` and it will pick up the fields automatically.

See ApplicationSyncRequest for an example.

jessesuen commented on server/users/users.proto:
```diff
@@ -0,0 +1,37 @@
+	rpc UpdatePassword(UsersPasswordRequest) returns (UserResponse) {
+		option (google.api.http) = {
+			put: "/api/v1/users/{name}/password"
+			body: "body"
```
body should be `*` after making suggested changes.

jessesuen commented on cmd/argocd/commands/users.go:
```diff
@@ -0,0 +1,71 @@
+			}
+
+			if CurrentPassword == "" {
+				CurrentPassword = settings.ReadAndConfirmPassword("Current Password")
```
We should not require that the user type in their current password twice. Password confirmation should be done only for the new one.

jessesuen commented on server/users/users.go:
```diff
@@ -0,0 +1,52 @@
+
+}
+
+//This Function is used to Update User Passwords \
```
Go doc convention is to repeat the name of the function in the comment. e.g.:

// UpdatePassword updates a user's password

jessesuen commented on util/settings/settings.go:
```diff
@@ -388,13 +388,13 @@ func (mgr *SettingsManager) notifySubscribers() {
+func ReadAndConfirmPassword(types string) string {
 	for {
-		fmt.Print("*** Enter an admin password: ")
+		fmt.Printf("*** Enter an %s password: ", types)
```
Lets just remove types argument and say "Enter password:" and "Confirm password"

jessesuen commented on server/users/users.proto:
```diff
@@ -0,0 +1,37 @@
+}
+message UsersPasswordRequest {
+    string name = 1;
+    Body body = 2;
```
Remove body and have currentPassword and newPassword as fields here.

jessesuen commented on server/users/users.proto:
```diff
@@ -0,0 +1,37 @@
+    string currentPassword = 2;
+    string newPassword = 3; 
+}
+message UsersPasswordRequest {
```
message name should be UpdatePasswordRequest

jessesuen commented on server/users/users.proto:
```diff
@@ -0,0 +1,37 @@
+    Body body = 2;
+}
+
+message UserResponse {}
```
message name should be UpdatePasswordResponse

JazminGonzalez-Rivero commented:
Hello @jessesuen 
1) changed the response/request message name
2)removed types argument from ReadAndConfirmPassword
3) current password only needs to be entered once 

but before i update the body, I wanted to explain why I did it this way. 

in the current setup the swagger UI shows the body and required path parameter that we would expect,

![image](https://user-images.githubusercontent.com/1447492/42725890-5d1fbfa8-8740-11e8-93a9-8ebf74b40963.png)

but when I use '*' I see that the name gets duplicated in the body of the request. 

![image](https://user-images.githubusercontent.com/1447492/42725905-82391456-8740-11e8-8eb5-973a2a18d4d3.png)

(also when I looked online, the documentation for using '*' mentioned "Note that when using * in the body mapping, it is not possible to have HTTP parameters, as all fields not bound by the path end in the body. This makes this option more rarely used in practice of defining REST APIs. The common usage of * is in custom methods which don't use the URL at all for transferring data." which lead me to searching for an alternative)

but let me know if this should still be changed, it's a quick update.

jessesuen commented:
Thanks for pointing out your reasoning.

> when I use '*' I see that the name gets duplicated in the body of the request.

Right, this is an unfortunate side affect of grpc-gateway. 

> as all fields not bound by the path end in the body.

This is the behavior we want, the only consequence is the name also gets into the body, despite already being in the path.

It's more important that we get the gRPC messages and data structures in a way that we like, then swagger docs. Having an extra Body field just to make the swagger docs cleaner, unnecessarily makes the gRPC API more confusing, which will be our preferred way of interacting with the API server.

jessesuen commented:
> I'll be in the mountains without connectivity until Monday night, but I can make the change then

Given that you'll be out, I'll go ahead and merge your current progress. I also realize we are missing an RBAC policy rule and check for user management, so I will pick off where you left off.

JazminGonzalez-Rivero commented:
Kk i have the change staged on my computer so i will push it first thing in the morning.
